composer.json

```diff
@@ -28,7 +28,7 @@
         }
     },
     "require": {
-        "php": "^7.1",
+        "php": "^7.3 || ^8.0",
```

Comment:
@boesing is there a record of the discussion about using ~8.0.0? That means when 8.1 comes out any of your dependencies will require a new release for you to be installable even if there are no code changes required. This feels a step backwards. I think the approach recently taken by Symfony to change their constraints to `>=` is the way to go but I can understand libraries choosing to go with `^8.0` - but ~8.0.0 feels very restrictive.

Reply:
>  That means when 8.1 comes out any of your dependencies will require a new release of their laminas dependency for them to be installable even if there are no code changes required to the laminas library.

This is intentional: PHP is too unstable to pin to a major release, and indeed, restricting is intentional. As for symfony, I can only say that jumping off the bridge because somebody told you to do it is not a wise approach to decision-making.

If all is good, next year, it is just a matter of bumping dependency ranges: until then, there is no guarantee that 8.1 will be compatible :-)

As for the meeting where this was decided, I think it was this one: https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-09-07-TSC-Minutes.md#hacktoberfest

Can't find the precise point in any meeting notes though - would need to check the `#tsc-meeting` channel on https://laminas.slack.com/

Reply:
@Ocramius I'm not jumping off a bridge because somebody told me too. I'm being stopped sending out scouts across a bridge because someone is unnecessarily putting a stop sign in the way.

There are a number of reasons why I think constraining the maximum version of PHP a library can run on is a not healthy for the ecosystem. Here are a some:
* Increases friction to early adopters from testing your code and creating PRs to fix it
* Increases friction to dependencies testing against new PHP versions. They have to jump through hoops and if nothing is broken they then require a release from Laminas in order to run tests. It makes the ecosystem form a long chain and, if you are on the end of the chain, you have a lot waiting to do.
* Something hand wavy about software freedoms, gophp5, and Postel's law. It just feels wrong.
* Composer's PHP constraints don't really work this way anyway... if there's something like `"platform": { "php": 7.4.99 }` in the root composer.json then, even though I'm running PHP 8, composer will still get laminas dependencies and not inform me a project is incompatible.

I can understand limiting to major versions ie something like `^7.0` (although I disagree with it) but `~8.0.0` increases the friction for laminas's dependencies due to a vague "PHP is too unstable to pin to a major release". I've worked on making Drupal 7.0, 7.1, 7.2, 7.4 and 8.0 compatible. It's a very large codebase and has many dependencies. PHP is generally the least of the problems. The biggest problem in maintaining compatibility against multiple PHP versions at the moment is PHPUnit.

For example, any dependency of laminas-feed that wants to be compatible with PHP 8.0 is currently waiting on https://github.com/laminas/laminas-feed/pull/28 - which is only about code quality - and the PHP 8.0 compatibility PR only stopped PHP from issuing a deprecation message (a good thing to be sure) BUT now in order to be PHP 8 compatible your dependencies have to upgrade from 2.12.x to 2.13.0 - when there is no new functionality. There's a similar story for laminas-escaper and laminas-diactoros.

Reply:
> The biggest problem in maintaining compatibility against multiple PHP versions at the moment is PHPUnit.

This! For e.g. the only reason PHP version contraint for this package needed to be updated to `^7.3` is because of PHPUnit.

I don't know in what utopia Sebastian lives where the whole PHP ecosystem gets updated as soon a PHP version reaches EOL.

Reply:
No one stops you from using composer v2 with `--ignore-platform-req=php`.
So what is the real pain you want to express?

We just want to specify every PHP version as tested. This can only work if we have active tests against that version. So every user of laminas can be 100% sure (as of PHP 8.0) that the code is tested under the current version the user executes `composer update/install`.

Imho, thats a huge promise to all laminas users and gives a maximum of "safety". Bugs can rise even with tested versions but at least it is compatible and no-one has to check if a vendor package of laminas/mezzio/laminas-api-tool is compatible with PHP 8.
Users are not beta-testers for new PHP versions but actually thats what symfony and most other libs are doing. They let their users beta-test a probably unstable version of their package. Imho, thats the point what we should talk about and not about "friction".

Reply:
@alexpott
> For example, any dependency of laminas-feed that wants to be compatible with PHP 8.0 is currently waiting on laminas/laminas-feed#28 - which is only about code quality

Unfortunately this is not entirely correct. This pull request is completed but [the commits needs a correct signed off](https://github.com/laminas/laminas-feed/pull/28/checks?check_run_id=1336868041) by the creator of the pull request.
But I hope that we can solve the problem on laminas-feed quickly.

Reply:
I've just found https://github.com/php-http/psr7-integration-tests/pull/45 and I'm more than a bit bemused. But to find the laminas project asking for PHP 8 compatibility on project because it is holding back the ecosystem is a bit much. The points made are exactly the points I'm making about needing compatibility before a release in order to be able to test before. I quote...
> bump
>
> RC2 just dropped, so PHP 8 is reaching relative stability. A number of PSR-7 providers are trying to ensure they are ready (e.g., Diactoros, Slim), but we are all currently blocked until this is merged.

Diactoros has been sitting on approved merge requests for days, and similarly for Feeds (who's PHP 8 compatibility release is blocked on a new integration with Psalm...). This means that unless Drupal makes exceptions to its release management policies it'll be June 2021 before we have an officially PHP 8 compatible release.

Also given the time its taken to get 8.0.0 compatibility done setting your constraint to ~8.0.0 does not fill your dependencies with confidence that you'll 8.1.0 compatible releases in place in time for your dependencies to be 8.1.0 compatible.

Reply:
These are all valid view points. However, there's an aspect that those _consuming_ the library are missing from those _maintaining_ the library: Using `^` constraints makes maintaining the library far harder.

How?

Because every time we get a new minor or major PHP version, we have to increase our test matrix. And every time we increase the test matrix, we generally have to revisit the PHPUnit versions we use and the syntax each uses to ensure that tests will actually run. This makes even simple contributions often more complex because new failures arise that were not present before; additionally, if the contributor does not have every version of PHP currently supported by the library on hand, they may not even know there are test failures until CI/CD runs... which, when the matrix gets large, take longer and longer to complete. And we all know that when we have to wait for results, we tend to get bored and move on.

Yes, in an **ideal** world, there would be no BC breaks or deprecations within a given minor release series, whether that's PHP or libraries you use. But in point of fact, these arise **all the time**. That's the origin of the "PHP is too unstable to pin to a major release'" statement. Every single minor release that comes out, we MUST do a cycle to test our libraries against it, because inevitably some new deprecation or change **does** lead to breakage, whether that's simply a test failing because of a new deprecation notice, or, worse, **users** identifying issues. When a user jumps to a new version and suddenly the library does not work, that immediately becomes **our** problem, but now there's urgency to it.

On top of that, there's the question of how long we support a given minor or major release of a library. If the library has versions 1.0, 1.1, 1.2, 2.0, 2.1, 3.0, and 3.1, and all of them have a minimum supported version of `^7.0`, does that mean we have to provide bugfixes and security fixes to _all_ of these branches? or just the latest minor releases of each major branch (e.g., 1.2, 2.1, and 3.1)? or current major and previous major?

By narrowing the window of PHP versions we support (via the `~` constraint, or bumping the minimum minor we support), we make the maintenance story far easier:

- We provide security patches to branches where the minimum supported PHP version is still supported by community PHP.
- We can have a smaller number of PHPUnit versions we have to be compatible with.
- There's a smaller test matrix we need to fulfill.
- When a new minor or major version of PHP is released, we create a new minor release that changes the PHP constraints, to signal that we have actually _tested_ it on that version, versus _assumed_ it would remain compatible.

Reply:
> we generally have to revisit the PHPUnit versions we use and the syntax each uses to ensure that tests will actually run.

Fortunately the world has changed... https://twitter.com/s_bergmann/status/1328197689029386240

Reply:
Finall, Hallelujah!